src/python/pants/backend/jvm/tasks/jvm_compile/jvm_compile.py

```diff
@@ -576,19 +577,62 @@ def _compile_vts(self, vts, target, sources, analysis_file, upstream_analysis, c
         if self.get_options().capture_classpath:
           self._record_compile_classpath(classpath, vts.targets, outdir)
 
-        # If compiling a plugin, don't try to use it on itself.
-        javac_plugins_to_exclude = (t.plugin for t in vts.targets if isinstance(t, JavacPlugin))
         try:
           self.compile(self._args, classpath, sources, outdir, upstream_analysis, analysis_file,
                        log_file, settings, fatal_warnings, zinc_file_manager,
-                       javac_plugins_to_exclude)
+                       self._get_plugin_map('javac', target),
+                       self._get_plugin_map('scalac', target))
         except TaskError:
           if self.get_options().suggest_missing_deps:
             logs = self._find_failed_compile_logs(compile_workunit)
             if logs:
               self._find_missing_deps('\n'.join([read_file(log) for log in logs]), target)
           raise
 
+  def _get_plugin_map(self, compiler, target):
+    """Returns a map of plugin to args, for the given compiler.
+
+    Only plugins that must actually be activated will be present as keys in the map.
```

Comment:
These two paragraphs are pretty rich in information. Would be great as a blurb on the docsite (and could then reference the document here).